File: backend/app/core/sockets.py
from typing import Dict, Any, List
from fastapi import WebSocket

class ConnectionManager:
    def __init__(self):
        self.active_connections: Dict[int, List[WebSocket]] = {}

    async def connect(self, session_id: int, websocket: WebSocket):
        await websocket.accept()
        if session_id not in self.active_connections:
            self.active_connections[session_id] = []
        self.active_connections[session_id].append(websocket)

    def disconnect(self, session_id: int, websocket: WebSocket):
        if session_id in self.active_connections:
            self.active_connections[session_id].remove(websocket)

    async def broadcast(self, session_id: int, message: Dict[str, Any]):
        if session_id in self.active_connections:
            for connection in self.active_connections[session_id]:
                await connection.send_json(message)

manager = ConnectionManager()

File: backend/services/sockets.js
class ConnectionManager {
  constructor() {
    this.activeConnections = new Map();
  }

  connect(sessionId, ws) {
    if (!this.activeConnections.has(sessionId)) {
      this.activeConnections.set(sessionId, new Set());
    }
    this.activeConnections.get(sessionId).add(ws);
    console.log(`WebSocket connected for session ${sessionId}`);
  }

  disconnect(sessionId, ws) {
    if (this.activeConnections.has(sessionId)) {
      this.activeConnections.get(sessionId).delete(ws);
      if (this.activeConnections.get(sessionId).size === 0) {
        this.activeConnections.delete(sessionId);
      }
    }
  }

  broadcast(sessionId, message) {
    if (this.activeConnections.has(sessionId)) {
      const payload = JSON.stringify(message);
      this.activeConnections.get(sessionId).forEach(ws => {
        if (ws.readyState === 1) { // OPEN
          ws.send(payload);
        }
      });
    }
  }
}

module.exports = new ConnectionManager();
